### PR TITLE
.Net: Added support for BinaryData for ImageContent.

### LINKED ISSUE #4781: .Net: Add overload ctor to ImageContent that takes an string "url"
When using the Vision models of AzureOpenAI and OpenAI, the API allows for an Image to be attached either as an URL or as a Data URI. In Semantic Kernel we do have `ImageContent` as a KernelContent-Component, but unfortunately, it takes only an `System.Uri` as the parameter for the object. 

I suggest to following addition to `ImageContent`:

### Proposed API
```c#
public sealed class ImageContent : KernelContent
{
    // existing, I think we should just change the type
    // this is a breaking change, although it don't think we should have both
    // but alternatively we could just add another Property 'DataUri' instead of changing the type
    public Uri? Uri { get; set; }
    // replaced by this:
    public string Uri { get; set; }

    // existing, though changed to call the new overload instead of base
    public ImageContent(
        Uri uri,
        string? modelId = null,
        object? innerContent = null,
        Encoding? encoding = null,
        IReadOnlyDictionary<string, object?>? metadata = null)
        : this(uri.ToString(), modelId, innerContent, encoding, metadata)
    {
        this.Uri = uri;
    }

    public ImageContent(
        string dataUri,
        string? modelId = null,
        object? innerContent = null,
        Encoding? encoding = null,
        IReadOnlyDictionary<string, object?>? metadata = null)
        : base(innerContent, modelId, metadata)
    {
        this.Uri = uri;
    }
}
``` 
This would allow us to use Data URIs with embedded images for the Vision APIs.

### Usage
```c#
var chatHistory = new ChatHistory();
var image = System.Convert.ToBase64String(imageBytes);
var imageDataLink = $"data:{imageContentType};base64,{image}";
chatHistory.Add(new ImageContent(imageDataLink));
var result = await chat.GetChatMessageContentAsync(chatHistory);
```

### Alternate Design
This would be a breaking Change, since we change the Type of the `Uri` property on `ImageContent`, so instead 
* we could add another property `DataUri` to `ImageContent`, but this would also mean that we have to check which of the two properties are used and prefer one other the other, if both are filled. 
* We could also just make a `private string field` that holds the DataUri and when the `Uri` ctor is used we just store the Uri as string in that field and only use the field on implementation site.  
* We could also make `ImageContent` take a ROS<byte>, byte[] or even just a Stream of the image and handle all the Base64Encoding in the `ImageContent` ctor (I think I would prefer a static Create method over ctor for this option, sonce we can than read the stream async), but this would mean, that we also need a second parameter with the image-ContentType and this would not allow for non base64 encoded data uris

I am happy to work on the implementation, since I have to do it anyway, because we need the solution fairly soon.

### PR DESCRIPTION
**Pull request by dersia**
Reopend PR. See https://github.com/microsoft/semantic-kernel/pull/4919 for previous PR. 
Reopening this PR from another feature branch makes rebasing easier.

### Motivation and Context

As Described in #4781 right now there is no possibility  in SK to add Images as DataUris to ChatCompletion APIs, although the Azure OpenAI API and the Open AI API both support this.

Fixes #4781

### Description

As per Discussion added overload to the ImageContent ctor that takes BinaryData.
For backward Compat we kept the ctor that takes an URI.

Also the new ctor throws, if the BinaryData is null, empty or if there is not MediaType provided.

I thought about allowing plain, non base64 encoded DataUris with BinaryData.
The Idea was to not encode to base64, if the MediaType is set to "text/plain", but then I decided, that this is not needed, since `Uri` in general allows for DataUris like `new Uri("data:text/plain;http://exmpaledomain.com")` just not for DataUris that are longer than 65520 bytes. I feel like that is ok, for plain DataUris.

We can still add this if needed.

Also as per discussion in the issue, I did not add additional overloads for direct Streams support.

### Contribution Checklist

- [x] The code builds clean without any errors or warnings
- [x] The PR follows the [SK Contribution Guidelines](https://github.com/microsoft/semantic-kernel/blob/main/CONTRIBUTING.md) and the [pre-submission formatting script](https://github.com/microsoft/semantic-kernel/blob/main/CONTRIBUTING.md#development-scripts) raises no violations
- [x] All unit tests pass, and I have added new tests where possible
- [x] I didn't break anyone :smile:

### CODE REVIEW

**Comment by dersia** [comment]
@RogerBarreto I fixed the comments. I also checked for the `using` pattern and I couldn't replicate the behavior I had in the past with the memory stream, so I reverted it back to using the `using` pattern in the test.

**Comment by dersia** [comment]
@Krzysztof318

> @dersia btw. new class was introduced to kernel [BinaryContent](https://github.com/microsoft/semantic-kernel/blob/main/dotnet/src/SemanticKernel.Abstractions/Contents/BinaryContent.cs) This class is currently used only by OpenAIFileService.
> 
> I know this news may not be relevant to this PR, but it is worth taking a view.

Interesting. Though I'm not sure I like the implementation, the Stream will advance if the `GetStreamAsync` or `GetContentAsync` is called and I also think that `Null`-Content/Stream would be fine and should not throw, it should just not be added to the request to the model.

**Comment by Krzysztof318** [comment]
@dersia I'm not thrilled with it either but I sent it to you as an overview. I think the way it is now is ok but it's best for @RogerBarreto to comment.

**Comment by RogerBarreto** [comment]
> I still think you should add information to the doc class comment that BinaryData takes precedence over URI.

@Krzysztof318 Thats a good point, @dersia you can add this as a `<remark></remark>` in the `ToString` `XmlDocs` comments.

**Review by RogerBarreto — APPROVED** [review]

File: dotnet/src/SemanticKernel.Abstractions/Contents/ImageContent.cs [hunk 1]
```diff
@@ -35,9 +40,47 @@ public sealed class ImageContent : KernelContent
...
+        }
+
+        this.Data = data;
+    }
+
+    /// <summary>
+    /// Returns the string representation of the image.
+    /// BinaryData images will be represented as DataUri
+    /// Remote Uri images will be represented as is
+    /// </summary>
```
Comment (RogerBarreto) [inline dotnet/src/SemanticKernel.Abstractions/Contents/ImageContent.cs:71]:
```suggestion
    /// </summary>
    /// <remarks>
    /// When Data is provided it takes precedence over URI
    /// </remarks>

```

**Review by RogerBarreto — APPROVED** [review]

**Review by dmytrostruk — APPROVED** [review]

**Pull request merged**